src/core/smtp/MCSMTPSession.cpp

```diff
@@ -864,6 +894,19 @@ void SMTPSession::noop(ErrorCode * pError)
     }
 }
 
+void SMTPSession::cancelMessageSending()
+{
+    // main thread
```

Comment:
mSmtp needs to be locked.

1/ We should perform a mailstream_cancel() only during mailesmtp_send().
2/ One issue is that mailesmtp_send_quit() will change mSmtp->stream.
Therefore, we need to make some changes in mailsmtp to add mailesmtp_send_quit_no_disconnect().

Here's the pseudo-code:

```
sendMessage() {
  lock()
  mCanCancel = true;
  unlock()
  mailesmtp_send()
  lock()
  mCanCancel = false;
  unlock()
}

cancelSend() {
  lock()
  if (mCanCancel) {
    mailstream_cancel()
  }
  unlock()
}
```